### PR TITLE
Update MaxTextTrainingEngine to work with Tunix's v2 orchestrator

### PR DESCRIPTION
**Pull request by igorts-git**
# Description

This PR is a squash of many commits from @A9isha and @khatwanimohit to enable the new RL flow with MaxTextTrainingEngine and Raiden weight sync. 

Disclaimer: This code is still WIP. It is not (yet) used by any production code. We need to merge it to unblock model experimentation while we continue to make many more polishing passes over it in the coming weeks.

# Tests

Verified live end-to-end: a 2-step GRPO run on a real 2-slice v5p-8 GKE cluster (trainer + tpu-inference RLVllmSampler rollout) completed both steps with real Raiden weight-sync rounds, checksum-verified between source (trainer) and destination (rollout).

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/training_engine/maxtext_engine.py [hunk 1]
```diff
@@ -263,48 +450,130 @@ def _update_kernel(self, state_pure, accumulated_grads, micro_step_count, mean_l
...
+    while scalars the optimizer allocates eagerly (adamw's `count`) carry a
+    SingleDeviceSharding on device 0. jit rejects that combination with "Received
+    incompatible devices for jitted computation", naming two unrelated arrays and no
+    cause. Normalizing the strays to a replicated NamedSharding puts every argument on the
+    same device set, while leaving genuinely sharded parameters untouched.
+    """
+    leaf_sharding = getattr(leaf, "sharding", None)
+    if isinstance(leaf_sharding, jax.sharding.NamedSharding) and leaf_sharding.mesh == self._mesh:
+      return leaf_sharding
+    return jax.sharding.NamedSharding(self._mesh, jax.sharding.PartitionSpec())
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply src/maxtext/training_engine/maxtext_engine.py:527]: applied suggestion

File: src/maxtext/training_engine/maxtext_engine.py [hunk 2]
```diff
@@ -263,48 +450,130 @@ def _update_kernel(self, state_pure, accumulated_grads, micro_step_count, mean_l
...
+    """
+    data_sharding = sharding.get_input_data_sharding(self._config, self._mesh)
+    data_spec = tuple(data_sharding.spec)
+
+    def leaf_sharding(leaf):
+      rank = jnp.ndim(leaf)
+      spec = list(data_spec[:rank])
+      if spec and spec[0] is not None and leaf.shape[0] % self._batch_axis_size(spec[0]):
+        spec[0] = None
+      return jax.sharding.NamedSharding(self._mesh, jax.sharding.PartitionSpec(*spec))
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply src/maxtext/training_engine/maxtext_engine.py:564]: applied suggestion

**Comment by igorts-git** [comment]
To avoid conflicts with the ongoing work by Mohit I am going to create a new branch from this one and create a new PR #4998. This PR is abandoned.
